src/java.base/share/classes/sun/security/provider/NamedKeyPairGenerator.java

```diff
@@ -0,0 +1,145 @@
+/// after the call. Together, this makes sure that the caller does not need to
+/// make any defensive copy on the input and output arrays. This requirement
+/// also applies to abstract methods defined in [NamedKEM] and [NamedSignature].
+///
+/// Also, an implementation must not keep any extra copy of a private key.
+/// For key generation, the only copy is the one returned in the
+/// [#implGenerateKeyPair] call. For all other methods, it must not make
+/// a copy of the input private key. A `KEM` implementation also must
+/// not keep a copy of the shared secret key, no matter if it's an
+/// encapsulator or a decapsulator.
+///
+/// The `NamedSignature` and `NamedKEM` classes provide `implCheckPublicKey`
+/// and `implCheckPrivateKey` methods that allow an implementation to validate
+/// a key before using it. An implementation may return a parsed key of
+/// a local type, and this parsed key will be passed to an operational method
+/// (For example, `implSign`) later. An implementation must not retain
+/// a reference of the parsed key.
+public abstract class NamedKeyPairGenerator extends KeyPairGeneratorSpi {
+
+    private final String fname; // family name
+    private final String[] pnames; // allowed parameter set name (at least one)
+
+    protected String name = null; // init as
+    private SecureRandom secureRandom;
+
+    /// Creates a new `NamedKeyPairGenerator` object.
+    ///
+    /// @param fname the family name
+    /// @param pnames supported parameter set names, at least one is needed.
+    ///     If multiple, the first one becomes the default parameter set name.
+    protected NamedKeyPairGenerator(String fname, String... pnames) {
+        this.fname = Objects.requireNonNull(fname);
+        if (pnames == null || pnames.length == 0) {
+            throw new AssertionError("pnames cannot be null or empty");
+        }
+        this.pnames = pnames;
+    }
+
+    private String checkName(String name) throws InvalidAlgorithmParameterException  {
+        for (var pname : pnames) {
+            if (pname.equalsIgnoreCase(name)) {
+                // return the stored standard name
+                return pname;
+            }
+        }
+        throw new InvalidAlgorithmParameterException(
+                "Unknown parameter set name: " + name);
+    }
+
+    @Override
+    public void initialize(AlgorithmParameterSpec params, SecureRandom random)
+            throws InvalidAlgorithmParameterException {
+        if (params instanceof NamedParameterSpec spec) {
+            name = checkName(spec.getName());
+        } else {
+            throw new InvalidAlgorithmParameterException(
+                    "Unknown AlgorithmParameterSpec: " + params);
+        }
+        this.secureRandom = random ;
+    }
+
+    @Override
+    public void initialize(int keysize, SecureRandom random) {
+        if (keysize != -1) {
+            // Bonus: a chance to provide a SecureRandom without
```

Comment:
Is this a TODO because I don't see the code that does this? I would avoid the word "Bonus" here.

Reply:
I just meant user can call `initialize(-1, rand)` to provide a user-defined `SecureRandom` without touching the parameter set used. Will update the comment.